junit-jupiter-api/src/main/java/org/junit/jupiter/api/DisplayNameGenerator.java

```diff
@@ -141,4 +168,131 @@ private static boolean hasParameters(Method method) {
+	 *
+	 * <p>This implements the functionality of {@link DisplayNameGenerator}
+	 * by generating complete sentences display names, these
+	 * sentences are divided with a separator, and the generator and separator
+	 * can be customisable by using the {@link IndicativeSentencesGeneration}
+	 * interface as annotation.
+	 *
+	 * @since 5.7
+	 */
+	@API(status = EXPERIMENTAL, since = "5.7")
+	class IndicativeSentences implements DisplayNameGenerator {
+
+		static final DisplayNameGenerator INSTANCE = new IndicativeSentences();
+
+		@Override
+		public String generateDisplayNameForClass(Class<?> testClass) {
+			return getGeneratorForIndicativeSentence(testClass).generateDisplayNameForClass(testClass);
+		}
+
+		@Override
+		public String generateDisplayNameForNestedClass(Class<?> nestedClass) {
+			return classReplaceToIndicativeSentence(nestedClass);
+		}
+
+		@Override
+		public String generateDisplayNameForMethod(Class<?> testClass, Method testMethod) {
+			return classReplaceToIndicativeSentence(testClass) + getSentenceSeparator(testClass)
+					+ getGeneratorForIndicativeSentence(testClass).generateDisplayNameForMethod(testClass, testMethod);
+		}
+
+		private String classReplaceToIndicativeSentence(Class<?> testClass) {
+			Class<?> enclosingParent = testClass.getEnclosingClass();
+			Optional<DisplayName> displayName = AnnotationSupport.findAnnotation(testClass, DisplayName.class);
+			Optional<DisplayNameGeneration> displayNameGeneration = AnnotationSupport.findAnnotation(testClass,
+				DisplayNameGeneration.class);
+
+			if (enclosingParent == null) {
+				return displayName.map(DisplayName::value).orElseGet(() -> generateDisplayNameForClass(testClass));
+			}
+			else {
+				if (displayNameGeneration.isPresent()) {
+					return displayName.map(DisplayName::value).orElseGet(() -> generateDisplayNameForClass(testClass));
+				}
+				else {
+					return displayName.map(name -> classReplaceToIndicativeSentence(enclosingParent)
+							+ getSentenceSeparator(testClass) + name.value()).orElseGet(
+								() -> classReplaceToIndicativeSentence(enclosingParent)
+										+ getSentenceSeparator(testClass) + getGeneratorForIndicativeSentence(
+											testClass).generateDisplayNameForNestedClass(testClass));
+				}
+			}
+		}
+
+		/**
+		 * Gets the separator for {@link IndicativeSentencesGeneration} when extracting the
+		 * annotation from {@code IndicativeSentencesGeneration}, if it doesn't find it,
+		 * then search for the parent classes, if no separator is found use @code{", "} by default.
+		 *
+		 * @param testClass Class to get Indicative sentence annotation separator either custom or default
+		 * @return the indicative sentence separator
+		 */
+		private String getSentenceSeparator(Class<?> testClass) {
+			Optional<IndicativeSentencesGeneration> indicativeSentencesGeneration = getIndicativeSentencesGeneration(
+				testClass);
+			if (indicativeSentencesGeneration.isPresent()) {
```

Comment:
Same kind of thought here, I would negate this `if` condition and get rid of the `else` block... but I guess it's a matter of taste really ^^

Reply:
Sure same as my last answer, thanks for the tips. 🙌